superset/reports/commands/execute.py

```diff
@@ -192,45 +200,173 @@ def _send(self, report_schedule: ReportSchedule) -> None:
+    - Error
+    """
+
+    current_states = [ReportState.NOOP, ReportState.ERROR]
+    initial = True
+
+    def next(self) -> None:
+        self.set_state_and_log(ReportState.WORKING)
+        try:
+            # If it's an alert check if the alert is triggered
+            if self._report_schedule.type == ReportScheduleType.ALERT:
+                if not AlertCommand(self._report_schedule).run():
+                    self.set_state_and_log(ReportState.NOOP)
+                    return
+            self._send()
+            self.set_state_and_log(ReportState.SUCCESS)
+        except CommandException as ex:
+            self.set_state_and_log(ReportState.ERROR, error_message=str(ex))
+            raise ex
+
+
+class ReportWorkingState(BaseReportState):
+    """
+    Handle Working state
+    next states:
+    - Error
+    - Working
+    """
+
+    current_states = [ReportState.WORKING]
+
+    def next(self) -> None:
+        if self.is_working_timeout():
+            exception_timeout = ReportScheduleWorkingTimeoutError()
+            self.set_state_and_log(
+                ReportState.ERROR, error_message=str(exception_timeout),
+            )
+            raise exception_timeout
+        exception_working = ReportSchedulePreviousWorkingError()
+        self.set_state_and_log(
+            ReportState.WORKING, error_message=str(exception_working),
+        )
+        raise exception_working
+
+
+class ReportSuccessState(BaseReportState):
+    """
+    Handle Success, Grace state
+    next states:
+    - Grace
+    - Not triggered
+    - Success
+    """
+
+    current_states = [ReportState.SUCCESS, ReportState.GRACE]
+
+    def next(self) -> None:
+        self.set_state_and_log(ReportState.WORKING)
+        if self._report_schedule.type == ReportScheduleType.ALERT:
+            if self.is_in_grace_period():
+                self.set_state_and_log(
+                    ReportState.GRACE,
+                    error_message=str(ReportScheduleAlertGracePeriodError()),
+                )
+                return
+            self.set_state_and_log(
+                ReportState.NOOP,
+                error_message=str(ReportScheduleAlertEndGracePeriodError()),
+            )
+            return
+        try:
+            self._send()
+            self.set_state_and_log(ReportState.SUCCESS)
+        except CommandException as ex:
+            self.set_state_and_log(ReportState.ERROR, error_message=str(ex))
+
+
+class ReportScheduleStateMachine:  # pylint: disable=too-few-public-methods
+    """
+    Simple state machine for Alerts/Reports states
+    """
+
+    states_cls = [ReportWorkingState, ReportNotTriggeredErrorState, ReportSuccessState]
+
+    def __init__(
+        self,
+        session: Session,
+        report_schedule: ReportSchedule,
+        scheduled_dttm: datetime,
+    ):
+        self._session = session
+        self._report_schedule = report_schedule
+        self._scheduled_dttm = scheduled_dttm
+
+    def run(self) -> None:
```

Comment:
It looks like this will silently swallow not-found states?

Reply:
good point, yet it would be very unexpected. Addressed it